tensorboard/plugins/graph/graph_util.py

```diff
@@ -14,152 +14,110 @@
-      from_proto: A TensorBoard GraphDef to copy contents from.
+      graph_defs: TensorBoard GraphDefs to merge.
 
     Returns:
-      to_proto
+      TensorBoard GraphDef that merges all graph_defs with unique prefixes.
 
     Raises:
-      ValueError in case any assumption about GraphDef is violated: A
-      GraphDef should have unique node, function, and gradient function
-      names. Also, when merging GraphDefs, they should have not have nodes,
-      functions, or gradient function mappings that share the name but details
-      do not match.
+      ValueError in case GraphDef versions mismatch.
     """
-    if from_proto.version != to_proto.version:
-        raise ValueError("Cannot combine GraphDefs of different versions.")
+    if len(graph_defs) == 1:
+        return graph_defs[0]
+    elif len(graph_defs) == 0:
+        return graph_pb2.GraphDef()
 
-    try:
-        _safe_copy_proto_list_values(
-            to_proto.node, from_proto.node, lambda n: n.name
-        )
-    except _ProtoListDuplicateKeyError as exc:
-        raise ValueError("A GraphDef contains non-unique node names: %s" % exc)
-    except _SameKeyDiffContentError as exc:
-        raise ValueError(
-            (
-                "Cannot combine GraphDefs because nodes share a name "
-                "but contents are different: %s"
-            )
-            % exc
-        )
-    try:
-        _safe_copy_proto_list_values(
-            to_proto.library.function,
-            from_proto.library.function,
-            lambda n: n.signature.name,
-        )
-    except _ProtoListDuplicateKeyError as exc:
-        raise ValueError(
-            "A GraphDef contains non-unique function names: %s" % exc
-        )
-    except _SameKeyDiffContentError as exc:
-        raise ValueError(
-            (
-                "Cannot combine GraphDefs because functions share a name "
-                "but are different: %s"
-            )
-            % exc
-        )
+    dst_graph_def = graph_pb2.GraphDef()
 
-    try:
-        _safe_copy_proto_list_values(
-            to_proto.library.gradient,
-            from_proto.library.gradient,
-            lambda g: g.gradient_func,
-        )
-    except _ProtoListDuplicateKeyError as exc:
-        raise ValueError(
-            "A GraphDef contains non-unique gradient function names: %s" % exc
-        )
-    except _SameKeyDiffContentError as exc:
-        raise ValueError(
-            (
-                "Cannot combine GraphDefs because gradients share a gradient_func name "
-                "but map to different functions: %s"
+    if graph_defs[0].versions.producer:
+        dst_graph_def.versions.CopyFrom(graph_defs[0].versions)
+
+    for index, graph_def in enumerate(graph_defs):
+        if dst_graph_def.versions.producer != graph_def.versions.producer:
+            raise ValueError("Cannot combine GraphDefs of different versions.")
+
+        mapped_graph_def = _prepend_names("graph_%d" % (index + 1), graph_def)
```

Comment:
You might consider a function signature more like

```python
def _add_with_prepended_names(prefix, graph_to_add, destination_graph)
```

That way we can avoid the overhead of constructing the temporary graphdef that's returned here, only to then copy its nodes and functions into the destination graph def.  Since proto python objects aren't immutable, the python code [has to do a deep copy](https://developers.google.com/protocol-buffers/docs/reference/python-generated#repeated-message-fields) each time we move the nodes around, so it would be nice not to do that twice when we could do it just once.